src/Tools/dotnet-gcdump/DotNetHeapDump/EventPipeDotNetHeapDumper.cs

```diff
@@ -0,0 +1,252 @@
+                            dumpComplete = true;
+                        }
+                    };
+
+                    gcDumpSession.Source.Clr.GCBulkNode += delegate (GCBulkNodeTraceData data)
+                    {
+                        if (data.ProcessID != processID)
+                        {
+                            return;
+                        }
+
+                        eventPipeDataPresent = true;
+
+                        if ((getElapsed() - lastEventPipeUpdate).TotalMilliseconds > 500)
+                        {
+                            log.WriteLine("{0,5:n1}s: Making GC Heap Progress...", getElapsed().TotalSeconds);
+                        }
+
+                        lastEventPipeUpdate = getElapsed();
+                    };
+
+                    if (memoryGraph != null)
+                    {
+                        dumper.SetupCallbacks(memoryGraph, gcDumpSession.Source, processID.ToString());
+                    }
+
+                    using (gcDumpSession)
+                    {
+                        log.WriteLine("{0,5:n1}s: Starting to process events", getElapsed().TotalSeconds);
+                        gcDumpSession.Source.Process();
+                        log.WriteLine("{0,5:n1}s: EventPipe Listener dying", getElapsed().TotalSeconds);
+                    }
+                });
+
+                for (; ; )
+                {
+                    if (ct.IsCancellationRequested)
+                    {
+                        log.WriteLine("{0,5:n1}s: Cancelling...", getElapsed().TotalSeconds);
+                        break;
+                    }
+
+                    if (readerTask.Wait(100))
+                    {
+                        break;
+                    }
+
+                    if (!eventPipeDataPresent && getElapsed().TotalSeconds > 5)      // Assume it started within 5 seconds.  
+                    {
+                        log.WriteLine("{0,5:n1}s: Assume no .NET Heap", getElapsed().TotalSeconds);
+                        break;
+                    }
+
+                    if (getElapsed().TotalSeconds > timeout)       // Time out after `timeout` seconds. defaults to 30s.
+                    {
+                        log.WriteLine("{0,5:n1}s: Timed out after {1} seconds", getElapsed().TotalSeconds, timeout);
+                        break;
+                    }
+
+                    if (dumpComplete)
+                    {
+                        break;
+                    }
+                }
+
+                log.WriteLine("{0,5:n1}s: Shutting down gcdump EventPipe session", getElapsed().TotalSeconds);
+                if (gcDumpSession != null)
+                {
+                    gcDumpSession.EndSession();
+                }
+                else
+                {
+                    throw new ApplicationException("EventPipe session for gcDumpSession failed");
```

Comment:
For practical purposes this is probably fine so I'm not going to complain if you leave it as-is, but if we needed to be rigorous there are problems lurking here:
a) If cancellation happens quickly enough gcDumpSession will be null by design and that shouldn't be an error.
b) gcDumpSession is being modified on one thread and read on another without any locks. In a race condition thread 1 could write a non-null value but here on thread 2 a null value could still be observed. This would lead to the exception being thrown instead of the correct EndSession() being called.
c) This exception appears to mask whatever the real issue was that caused gcDumpSession to be null. Ideally the code would wait on readerTask to ensure that any exception thrown from readerTask is what propagates here.

Reply:
I refactored this a bit to remove some of the multi-threading and hiding of errors.  I'll ping again for another look.  The fixes got rid of a race that I found while refactoring.